src/builder/QuarkBuilder.sol

```diff
@@ -218,40 +226,51 @@ contract QuarkBuilder {
-                            chainAccountsList[i].chainId, transferIntent.chainId, transferIntent.assetSymbol
-                        )
-                ) {
-                    aggregateTransferAssetAvailableBalance += Accounts.sumBalances(positions);
-                    // If the payment token is the transfer token and user opt for paying with the payment token, reduce the available balance by the maxCost
-                    if (payment.isToken && Strings.stringEqIgnoreCase(payment.currency, transferIntent.assetSymbol)) {
-                        uint256 maxCost = PaymentInfo.findMaxCost(payment, chainAccountsList[i].chainId);
-                        aggregateTransferAssetAvailableBalance -= maxCost;
-                    }
-                }
+        // If no funds need to be bridged, then this check is satisfied
+        // TODO: We might still need to check the availability of funds on the target chain, e.g. see if
+        // funds are locked in a lending protocol and can't be withdrawn
+        if (!needsBridgedFunds(transferIntent, chainAccountsList)) {
+            return;
+        }
+
+        // Check each chain to see if there are enough transfer assets to be bridged over
+        uint256 aggregateTransferAssetAvailableBalance;
+        for (uint256 i = 0; i < chainAccountsList.length; ++i) {
+            Accounts.AssetPositions memory positions =
+                Accounts.findAssetPositions(transferIntent.assetSymbol, chainAccountsList[i].assetPositionsList);
+
+            // Skip the chain if action is paid for with token and max cost is not specified for the current chain
+            if (payment.isToken && !PaymentInfo.hasMaxCostForChain(payment, chainAccountsList[i].chainId)) {
+                continue;
             }
-            if (aggregateTransferAssetAvailableBalance < transferIntent.amount) {
-                revert FundsUnavailable();
+
+            if (
+                chainAccountsList[i].chainId == transferIntent.chainId
+                    || BridgeRoutes.canBridge(
+                        chainAccountsList[i].chainId, transferIntent.chainId, transferIntent.assetSymbol
+                    )
+            ) {
+                aggregateTransferAssetAvailableBalance += Accounts.sumBalances(positions);
+                // If the user opts for paying with the payment token and the payment token is the transfer token, reduce
+                // the available balance by the max cost because the max cost is reserved for paying the txn
+                if (payment.isToken && Strings.stringEqIgnoreCase(payment.currency, transferIntent.assetSymbol)) {
+                    uint256 maxCost = PaymentInfo.findMaxCost(payment, chainAccountsList[i].chainId);
+                    aggregateTransferAssetAvailableBalance -= maxCost;
+                }
             }
         }
+        if (aggregateTransferAssetAvailableBalance < transferIntent.amount) {
+            revert FundsUnavailable();
+        }
     }
```

Comment:
The diff here looks big, but most of the diff is just removing one layer of nesting by moving the outer check (`if needsBridgedFunds`) out. The real diff is adding this check:

```
// Skip the chain if action is paid for with token and max cost is not specified for the current chain
if (payment.isToken && !PaymentInfo.hasMaxCostForChain(payment, chainAccountsList[i].chainId)) {
    continue;
 }
```